Support for multiple FHIR profile extensions

chandrashekar-s opened the pull request:
## Description of what I changed

Fixes #980 

This is an extension to the PR #924. In this PR we provide support for multiple FHIR profile extension for a given resource type. This is achieved by creating a Master AvroConverter for the given resource type by merging all the individual AvroConverters against the configured extensions profiles including the base profile. The merging includes a deep merge of the nested elements under the AvroConverter.

## E2E test

Tested the changes with the default US Core FHIR profile extensions and verified that the extension fields are properly populated in the parquet files.

TESTED:

Please replace this with a description of how you tested your PR beyond the
automated e2e/unit tests.

## Checklist: I completed these to help reviewers :)

<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->

- [x] I have read and will follow the [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review [Java](https://google.github.io/styleguide/javaguide.html) and [Python](https://google.github.io/styleguide/pyguide.html) style guides.

- [x] My IDE is configured to follow the Google [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? -> [configure your IDE](https://github.com/google/google-java-format).

- [ ] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

chandrashekar-s commented:
@bashir2 The PR is ready for the first round of review, can you please have a look and let me know if the approach taken is good. In the meanwhile, I will add unit tests.

bashir2 commented on bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/converters/DefinitionToAvroVisitor.java:
```diff
@@ -702,60 +892,13 @@ public HapiConverter<Schema> visitMultiValued(
   public HapiConverter<Schema> visitChoice(
       String elementName, Map<String, HapiConverter<Schema>> choiceTypes) {
 
-    List<Field> fields =
```
I have not looked at the rest of the changes in this file yet (will do later); are we making any logic change here or is it just refactoring code that was here into the methods below?

bashir2 reviewed (commented):
Thanks @chandrashekar-s for the changes. I think this is a nice solution for the multiple extension issue and I like the `merge()` idea. I am sharing my first batch of comments. I have not yet finished my first pass and may share another batch later today, if I can (otherwise tomorrow).

chandrashekar-s commented on bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/AvroConverter.java:
```diff
@@ -196,4 +221,31 @@ public Schema getSchema() {
+    HapiConverter<Schema> mergedConverter =
+        left.hapiToAvroConverter.merge(right.hapiToAvroConverter);
+    RuntimeResourceDefinition[] resources = new RuntimeResourceDefinition[1];
+    resources[0] = context.getResourceDefinition(mergedConverter.getElementType());
```
I also added few tests in `AvroConverterMergeTest` to verify if the extensions are converted properly.

chandrashekar-s commented on bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/AvroConverter.java:
```diff
@@ -196,4 +221,31 @@ public Schema getSchema() {
+        left.hapiToAvroConverter.merge(right.hapiToAvroConverter);
+    RuntimeResourceDefinition[] resources = new RuntimeResourceDefinition[1];
+    resources[0] = context.getResourceDefinition(mergedConverter.getElementType());
+    return new AvroConverter(mergedConverter, resources);
```
Few tests are added in `AvroConverterMergeTest` to verify if the conversions from Hapi->Avro->Hapi is happening. I verified this for `extension` fields. But there are some issues while checking for other field using the `equalsDeep` method.

chandrashekar-s replied:
I have skipped the verification for the `Id` field and another field called `Text` during comparison. Created this [issue](https://github.com/google/fhir-data-pipes/issues/1003) for the ID inconsistency.

chandrashekar-s reviewed (commented):
Thanks for the review @bashir2. In the latest commits, I have added more unit tests and have addressed your comments.

chandrashekar-s commented on bunsen/bunsen-core/src/test/java/com/cerner/bunsen/ProfileMapperProviderTest.java:
```diff
@@ -19,38 +10,39 @@ public class ProfileMapperProviderTest {
-    assertThat(profileMapping.get("Observation"), Matchers.notNullValue());
-    assertThat(profileMapping.get("Observation"), Matchers.equalTo(US_CORE_OBSERVATION_RESULTS));
-  }
+  //
```
These tests have been updated in the latest commit.

chandrashekar-s commented on pipelines/common/src/test/java/com/google/fhir/analytics/AvroConversionUtilTest.java:
```diff
@@ -165,32 +168,34 @@ public void convertObservationWithBigDecimalValue() throws IOException, ProfileM
-    assertThat(
-        obsUtilAvroConverter.getSchema(), Matchers.equalTo(obsDirectAvroConverter.getSchema()));
-  }
+  //  @Test
```
These tests have been updated in the latest commit.

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/ParquetMerger.java:
```diff
@@ -245,7 +246,8 @@ public void processElement(ProcessContext c) {
 
-  public static void main(String[] args) throws IOException, ProfileMapperException {
+  public static void main(String[] args)
+      throws IOException, ProfileMapperException, HapiMergeException {
```
In continuation to the previous comment, I tried converting the avro record creating using the old schema O into a HAPI object using the new Schema, but it failed with the below error

```
java.lang.ArrayIndexOutOfBoundsException: Index 22 out of bounds for length 22

	at org.apache.avro.generic.GenericData$Record.get(GenericData.java:289)
	at com.cerner.bunsen.avro.converters.DefinitionToAvroVisitor$CompositeToAvroConverter.getChild(DefinitionToAvroVisitor.java:205)
```

chandrashekar-s commented on bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/converters/DefinitionToAvroVisitor.java:
```diff
@@ -200,6 +252,84 @@ protected boolean isMultiValued(Schema schema) {
+
+      CompositeToAvroConverter otherConverter = (CompositeToAvroConverter) other;
+      // For extensions
+      if (!(Strings.isNullOrEmpty(this.extensionUrl())
```
The new method `merge` for different classes gets tested automatically with the existing unit test cases.

chandrashekar-s replied:
Yes, this test cases have been already added.

chandrashekar-s commented:
@bashir2 Please have a look at the latest commits containing more unit tests and some minor fixes.

bashir2 reviewed (commented):
Thanks @chandrashekar-s for the changes. Most of my comments are minor and I think we can close this PR this week. Beside the comments below, there is also one unresolved comment about `ParquetMerger` which I cannot reply to anymore (I guess because you have force pushed over an old commit); it is about the experiments you did with compatible but different schemas (e.g., when a new extension is added). Can you please document your finding somewhere? Even if it is in a javadoc that is fine, but it would be nice to have a record of that.

chandrashekar-s commented on bunsen/bunsen-avro/src/test/java/com/cerner/bunsen/avro/AvroConverterMergeTest.java:
```diff
@@ -139,7 +141,11 @@ public void validateMergedStu3UsCoreResourceWithExtensions()
-    validateStu3Extensions(patient.getExtension(), patientDecoded.getExtension());
+    patientDecoded.setId(patient.getId());
+    // TODO : The test field is not properly copied to the decoded object back, hence manually
+    // copying it. This needs to be debugged.
```
This is different from the ID issue. Here, due to some reason the `text` field is not properly being copied to `Avro` and hence is lost when converted back to `Hapi` again. I will file a new issue for this.

chandrashekar-s replied:
Have raised this [issue](https://github.com/google/fhir-data-pipes/issues/1014) and updated the comments.

chandrashekar-s commented on bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/converters/DefinitionToAvroVisitor.java:
```diff
@@ -254,20 +225,14 @@ protected boolean isMultiValued(Schema schema) {
       // For extensions
       if (!(Strings.isNullOrEmpty(this.extensionUrl())
           && Strings.isNullOrEmpty(otherConverter.extensionUrl()))) {
         if (!Objects.equals(this.extensionUrl(), otherConverter.extensionUrl())) {
```
Yes, I had thought about this. The reason I could not move all the validation logic to `HapiConverterUtil.validateIfConvertersCanBeMerged` is that some of the implementations of `HapiConverter` are private classes in `DefinitionToAvroVisitor` and hence I couldn't access them in `HapiConverterUtil`. I will move back the validations to each specific implementations now.

chandrashekar-s replied:
I have refactored this piece to have validations in respective implementationm classes. However, to increase reusability of code, I have still retained some bits of code in `HapiConverterUtil` and moved some common code to `PrimitiveConverter`.

chandrashekar-s replied:
Changed this in the latest commits, validations have been moved to respective classes.

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/HapiConverter.java:
```diff
@@ -83,6 +84,15 @@ public String getElementType() {
   }
 
+  /**
+   * Returns the elements which the converter if responsible for conversion
```
This is not relevant anymore, as the method is removed.

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/HapiConverter.java:
```diff
@@ -83,6 +84,15 @@ public String getElementType() {
+   *
+   * @return the elements which needs conversion
+   */
+  public Object getElements() {
```
This is a good suggestion. I have made the changes according to the suggestions which you have given. Thanks!

chandrashekar-s commented on pipelines/common/src/test/java/com/google/fhir/analytics/AvroConversionUtilTest.java:
```diff
@@ -166,31 +163,34 @@ public void convertObservationWithBigDecimalValue() throws IOException, ProfileM
     assertThat(
-        patientUtilAvroConverter.getSchema(),
-        Matchers.equalTo(patientDirectAvroConverter.getSchema()));
+        patientConverterFetchedDirectly.getSchema().toString(),
```
Yes, you are right, there is no need to convert the schema to string for comparison. I am not sure why I had changed to `toString()`. Reverted back now.

chandrashekar-s commented on bunsen/bunsen-avro/src/test/java/com/cerner/bunsen/avro/AvroConverterMergeTest.java:
```diff
@@ -0,0 +1,167 @@
+        ProfileMapperFhirContexts.getInstance()
+            .contextFromClasspathFor(FhirVersionEnum.R4, "/r4-us-core-definitions");
+    validateSchema(
+        "/r4-us-core-schemas/us-core-patient-schema.json",
```
I had generated the schema files using the `AvroConverter.getSchema()` for the base resources and manually made sure for extensions the new schema has the additional fields by again generating using the same method `AvroConverter.getSchema()`. We can discuss on this tomorrow.

chandrashekar-s replied:
Documentation is done on how the schema files are generated in a README.md file in the directory `src/test/resources`

chandrashekar-s commented:
Thanks @bashir2 for the detailed review. I have addressed most of the comments, there are only 2 comments left which we can discuss in our tomorrow's meeting.

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/StructureDefinitions.java:
```diff
@@ -280,7 +280,10 @@ private <T> List<StructureField<T>> elementToFields(
+    } else if (element.hasSingleType()
+        && PRIMITIVE_TYPES.contains(element.getFirstTypeCode())
+        // Allow for choice types only if they are not Extensions
+        && (element.getPath().startsWith("Extension") || !element.getPath().endsWith("[x]"))) {
```
This is little tricky, as this piece of code needs to be executed for Extensions with a single choice type and not for non extensions. We can discuss on this.

bashir2 replied:
So as we discussed, let's update the comment above and also make sure the `"Extension"` constraint is required (maybe mention a particular extension SD in the comment where this is needed).

chandrashekar-s replied:
Comment is added on why the additional condition is needed

bashir2 reviewed (approved):
Thanks @chandrashekar-s for the changes; all remaining comments are minor and/or we discussed them. So please feel free to merge this once they are resolved.

chandrashekar-s commented on bunsen/bunsen-avro/src/test/java/com/cerner/bunsen/avro/AvroConverterMergeTest.java:
```diff
@@ -0,0 +1,167 @@
+        this.getClass().getResourceAsStream("/r4-custom-schemas/bunsen-test-patient-schema.json");
+    Schema expectedSchema = new Parser().parse(inputStream);
+
+    Assert.assertEquals(expectedSchema.toString(), mergedConverter.getSchema().toString());
```
Done

chandrashekar-s reviewed (commented):
Thanks @bashir2 for the review. I have addressed the comments in the latest commits. I will be merging the changes. Please let me know if there are any concerns.
